fairlearn/metrics/_metric_frame.py

```diff
@@ -0,0 +1,515 @@
+
+        Returns
+        -------
+        pandas.DataFrame
+            The result of evaluating
+            each underlying metric for each combination of classes in the
+            sensitive and control features. The columns identify
+            the underlying metric, while the rows are indexed by the unique
+            combinations of classes found in the sensitive and control
+            features. If control features are present, they are in the
+            'outer' layer of the row indexing.
+
+            If a particular combination of classes was not present in the dataset
+            (likely to occur as more sensitive and control features
+            are specified), then the corresponding entry in the DataFrame will
+            be NaN.
+        """
+        return self._by_group
+
+    @property
+    def control_levels(self) -> List[str]:
+        """Return a list of feature names which are produced by control features.
+
+        If control features are present, then the rows of the ``by_group``
+        property have a :class:`pd.MultiIndex` index. This property
+        identifies which elements of that index are control features.
+
+        Returns
+        -------
+        List[str]
+            List of names, which can be used in calls to
+            :meth:`pandas.DataFrame.groupby` etc.
+        """
+        return self._cf_names
+
+    @property
+    def sensitive_levels(self) -> List[str]:
+        """Return a list of the feature names which are produced by sensitive features.
+
+        In cases where the ``by_group`` property has a :class:`pd.MultiIndex`
+        index, this identifies which elements of the index are sensitive features.
+
+        Returns
+        -------
+        List[str]
+            List of names, which can be used in calls to
+            :meth:`pandas.DataFrame.groupby` etc.
+        """
+        return self._sf_names
+
+    def group_max(self) -> Union[pd.Series, pd.DataFrame]:
+        """Return the maximum value of the metric over the sensitive features.
+
+        This method computes the maximum value over all combinations of
+        sensitive features for each underlying metric function in the ``by_groups``
+        property (it will only succeed if all the underlying metric
+        functions return scalar values). If there are no control
+        features, then the result will be a :class:`pandas.Series`,
+        with rows corresponding to the underying metrics.
+
+        Returns
+        -------
+        pandas.Series or pandas.DataFrame
+            If there are control features, then the result is a
+            :class:`pandas.DataFrame` with columns corresponding to the
+            underlying metrics, and rows indexed by the classes identified
+            by the control features. Otherwise, a :class:`pandas.Series`
+            indexed by the names of the underlying metrics.
+        """
+        if not self.control_levels:
```

Comment:
this should simplify to `return self.by_group.max(level=control_levels)` once we update the default `control_levels`.

Reply:
Some simplifications done, but internals can wait on a refactor.